benchmarking-a-deployment.md

```diff
@@ -0,0 +1,212 @@
+2019/01/21 11:15:25 Removing container bh2qj2vo3v1lt1ft4nrg ...
+2019/01/21 11:15:25 Removing container bh2qj0fo3v1lt1ft4nqg ...
+2019/01/21 11:15:25 Removing container bh2qicfo3v1lt1ft4nig ...
+2019/01/21 11:15:25 Removing container bh2qihfo3v1lt1ft4nkg ...
+2019/01/21 11:15:25 Removing container bh2qijvo3v1lt1ft4nlg ...
+2019/01/21 11:15:25 Removing container bh2qimfo3v1lt1ft4nmg ...
+2019/01/21 11:15:25 Removing container bh2qiovo3v1lt1ft4nng ...
+2019/01/21 11:15:25 Removing container bh2qi4vo3v1lt1ft4nfg ...
+2019/01/21 11:15:25 Removing container bh2qirfo3v1lt1ft4nog ...
+2019/01/21 11:15:25 Removing container bh2qhvvo3v1lt1ft4ndg ...
+2019/01/21 11:15:25 Removing container bh2qi9vo3v1lt1ft4nhg ...
+2019/01/21 11:15:39 Containers boot time measurement:
+2019/01/21 11:15:39   Launching all 15 containers took 2m36.560310342s
+2019/01/21 11:15:39   Out of 15 containers 0 failed to launch
+2019/01/21 11:15:39   Average container launch time: 6.149119411s
+2019/01/21 11:15:39   Max container launch time: 6.576302043s
+2019/01/21 11:15:39   Min container launch time: 5.911184959s
+2019/01/21 11:15:39   Android system failed to boot in the following containers:
+2019/01/21 11:15:39     None
+2019/01/21 11:15:39 Containers statistics:
+2019/01/21 11:15:39   FPS avg 58 min 52 max 64 for 15 containers
+2019/01/21 11:15:39   Containers below run at low FPS(<30):
+2019/01/21 11:15:39     None
+```
+
+This launches 15 containers on Anbox null platform with default display specification
+
+ | Display Spec            | Value |
+ | ----------------------- | ----- |
+ | Display width           | 1280  |
+ | Display height          | 720   |
+ | FPS                     | 60    |
+ | Display density         | 160   |
+
+for benchmarking.
+
+A different display specification can be configured through `--user-data` parameter when running the benchmark. The content format of user-data varies based on different platforms.
+
+| Platform              | Content format of user data           | Example                                                   |
+| --------------------- | ------------------------------------  | --------------------------------------------------------  |
+| Null                  | Comma separated values                | <display_width>,<display_height>,<display_fps>,<display_density>  |
+| Swrast                | Comma separated values                | <display_width>,<display_height>,<display_fps>,<display_density>  |
+| WebRTC                | Json-based                            | {<br>"display_width": <display_width>,<br>"display_height": <display_height>,<br>"display_density": <display_density>,<br>"fps": <display_fps>,<br>"render_only": true<br> } |
+
+> **Note:** when running benchmark against webrtc platform, Anbox session must be launched with render only mode since `amc benchmark` doesn't require WebRTC negotiation to establish WebRTC connection upon container creation.
```

Comment:
```suggestion
> **Note:** When running a benchmark against the `webrtc` platform, you must launch the Anbox session in render-only mode. The `amc benchmark` command doesn't require WebRTC negotiation to establish a WebRTC connection upon container creation.
```
Is it correct that the user launches the session?
And is it really a "must"? The explanation sounds to me like render-only mode is sufficient, but you could also use other modes.

Reply:
Is it correct that the user launches the session?
A:  Whenever creating a container, internally we launch a fresh Anbox session, which talks to Anbox container manager to boot up Android system. Perhaps to make it easy to understand, we may stick to **launch a container** ?
      
And is it really a "must"? The explanation sounds to me like render-only mode is sufficient, but you could also use other modes.
A: yes, it is. otherwise, creating containers would fail immediately right after they just got started due to the missing session URL since the `amc benchmark` command doesn't interact with stream gateway, hence no session URL is given at all.  

I am not quite sure if we want to leak too many implementation details about this. What about 

> **Note:** Because the `amc benchmark` command doesn't interact with stream gateway for the benchmark execution, when running a benchmark against the `webrtc` platform, you must launch containers in render-only mode, otherwise container creation would fail during the benchmark runtime.

Reply:
But then it is the benchmark command that launches the containers, and not the user (at least not directly)?

Maybe:
```suggestion
> **Note:** Because the `amc benchmark` command doesn't interact with the stream gateway for the benchmark execution, the containers are launched in render-only mode when running a benchmark against the `webrtc` platform. Otherwise, the container creation would fail.
```

Reply:
But then it is the benchmark command that launches the containers, and not the user (at least not directly)?
A: yes, it's the benchmark command that launches the containers, however the display spec is specified by a user, not benchmark command itself. We make it clear that people have to declare the `render-only: true` in the display specs against webrtc platform.  See the json-based content in the display spec table. Things can easily mess up if `render-only: true` is not present in the display spec, which is going to be specified by end user.

Reply:
Ah, right - now I get it. I missed the render-only in the table. ;)

Then how about:
```suggestion
> > **Note:** If you're running a benchmark against the `webrtc` platform, make sure to specify `"render_only": true` to launch the containers in render-only mode. Otherwise, the container creation will fail, because the `amc benchmark` command doesn't interact with the stream gateway for the benchmark execution.
```

Reply:
+1, much better.